src/workerd/jsg/modules.c++

```diff
@@ -4,10 +4,35 @@
 
 #include "modules.h"
 #include "promise.h"
+#include <kj/mutex.h>
 
 namespace workerd::jsg {
 namespace {
 
+class CompileCache {
+  // The CompileCache is used to hold cached compilation data for built-in JavaScript modules.
+public:
+  void add(kj::StringPtr key, std::unique_ptr<v8::ScriptCompiler::CachedData> cached) {
+    cache.lockExclusive()->insert(kj::str(key), kj::mv(cached));
+  }
+
+  kj::Maybe<v8::ScriptCompiler::CachedData&> find(kj::StringPtr key) {
+    return cache.lockShared()->find(key).map([](auto& data)
+        -> v8::ScriptCompiler::CachedData& { return *data; });
+  }
+
+private:
+  kj::MutexGuarded<kj::HashMap<kj::String, std::unique_ptr<v8::ScriptCompiler::CachedData>>> cache;
+  // TODO(review): There's a question here: Is it better to have a single CompileCache for the
+  // entire process, shared by all threads, or individual per-thread caches? Per-thread caches
+  // will use more memory but would avoid the locking here. However, the locking shouldn't be too
+  // bad while we have a relatively small number of builtins... and it's entirely likely that the
+  // cache will be populated very quickly in the process and shouldn't change much after, so
+  // exclusive locks should be rare.
+};
+
+CompileCache compileCache;
```

Comment:
Can we do:

```
const CompileCache& getCompileCache() {
  static CompileCache instance;
  return instance;
}
```

This way, the constructor runs the first time the function is called, rather than at startup. Startup-time constructors have all kinds of problems and KJ style insists on completely avoiding them outside of tests. (The Cap'n Proto build actually makes them an error, but I guess that flag didn't carry over to the Workers build.)

Additionally, KJ style prefers to declare objects `const` when they are accessed from multiple threads, so that we can use const-correctness to enforce thread-safety. Your code above uses a mutex, so the methods can both be declared `const`.